src/ui/Controls/AudioVisualizer.cs

```diff
@@ -431,40 +431,52 @@ private void LoadParagraphs(Subtitle subtitle, int primarySelectedIndex, ListVie
             const double additionalSeconds = 15.0; // Helps when scrolling
             var startThresholdMilliseconds = (_startPositionSeconds - additionalSeconds) * TimeCode.BaseUnit;
             var endThresholdMilliseconds = (EndPositionSeconds + additionalSeconds) * TimeCode.BaseUnit;
-            var displayableParagraphs = new List<Paragraph>();
-            for (var i = 0; i < subtitle.Paragraphs.Count; i++)
-            {
-                var p = subtitle.Paragraphs[i];
-
-                if (p.StartTime.IsMaxTime)
-                {
-                    continue;
-                }
 
-                _subtitle.Paragraphs.Add(p);
-                if (p.EndTime.TotalMilliseconds >= startThresholdMilliseconds && p.StartTime.TotalMilliseconds <= endThresholdMilliseconds)
-                {
-                    displayableParagraphs.Add(p);
-                    if (displayableParagraphs.Count > 99)
-                    {
-                        break;
-                    }
-                }
+            List<Paragraph> thresholdParagraphs = subtitle.Paragraphs.Where(p => !p.StartTime.IsMaxTime)
+                .Where(p => p.EndTime.TotalMilliseconds >= startThresholdMilliseconds && p.StartTime.TotalMilliseconds <= endThresholdMilliseconds).ToList();
+            _subtitle.Paragraphs.AddRange(thresholdParagraphs);
+
+            double startVisibleMilliseconds = _startPositionSeconds * TimeCode.BaseUnit;
+            double endVisibleMilliseconds = EndPositionSeconds * TimeCode.BaseUnit;
+            List<Paragraph> visibleParagraphs = thresholdParagraphs.Where(
+                p => p.EndTime.TotalMilliseconds >= startVisibleMilliseconds && p.StartTime.TotalMilliseconds <= endVisibleMilliseconds).ToList();
+            List<Paragraph> invisibleParagraphs = thresholdParagraphs.Except(visibleParagraphs).ToList();
+
+            const int maxDisplayableParagraphs = 100;
+            IEnumerable<Paragraph> displayableParagraphs = visibleParagraphs;
+            if (visibleParagraphs.Count > maxDisplayableParagraphs)
+            {
+                /*
+                 * Group & select is done so that it draws paragraphs across the entire timeline before drawing overlapping paragraphs, up to the display limit.
+                 * Materialize to list so that it can be counted below without pruning twice.
+                 */
+                displayableParagraphs = visibleParagraphs.Where(p => p.DurationTotalMilliseconds >= 0.01)
```

Comment:
Too many LINQ. This can impact the performance!

Reply:
Thanks for taking a look! I didn't think the performance impact was too large with LINQ but I don't mind rewriting it to use traditional loops.

After creating this PR I noticed there's a SecondsToXPosition method that looks better than grouping by the same number of milliseconds no matter the zoom level. So I'll get both of these changes in this PR.

Reply:
@niksedk is this something of your interest?